src/Altinn.Platform/Altinn.Platform.Authentication/Altinn.Platform.Authentication.Tests/Controllers/AuthenticationControllerTests.cs

```diff
@@ -498,11 +503,335 @@ public async Task AuthenticateUser_NoTokenPortalParametersIncluded_RedirectsToSB
+        [Fact]
+        public async Task AuthenticateUserWithOIDC_InvalidState_BadRequest()
+        {
+            // Arrange         
+            string gotoUrl = "http://ttd.apps.localhost/ttd/testapp";
+            HttpClient client = GetTestClient(_cookieDecryptionService.Object, _userProfileService.Object, true, true);
+            string redirectUri = "http://localhost/authentication/api/v1/authentication?goto=" + HttpUtility.UrlEncode(gotoUrl);
+
+            string url = "/authentication/api/v1/authentication?goto=" + HttpUtility.UrlEncode(gotoUrl) + "&DontChooseReportee=true";
+            HttpRequestMessage redirectToOidcProviderRequest = new HttpRequestMessage(HttpMethod.Get, url);
+
+            // Call Authentication component equalt to browsr beeing redirect from Altinn app
+            HttpResponseMessage redirectToOidcProviderResponse = await client.SendAsync(redirectToOidcProviderRequest);
+
+            // Assert that user is redirected to correct Oidc provider and a XSRF Cookie was set
+            Assert.Equal(HttpStatusCode.Redirect, redirectToOidcProviderResponse.StatusCode);
+            Uri redirectToOidcProviderUri = new Uri(redirectToOidcProviderResponse.Headers.Location.ToString());
+            Assert.Equal("idprovider.azurewebsites.net/authorize", redirectToOidcProviderUri.Host + redirectToOidcProviderUri.AbsolutePath);
+
+            // Verify that XSRF token cookie set is set. 
+            ValidateXSRFTokenPresent(redirectToOidcProviderResponse);
+
+            // Verify that all required OIDC Params are set and have the correct values
+            ValidateOidcParams(redirectToOidcProviderUri, redirectUri, "2314534634r2", out string stateParam, out string nonceParam, out string redirectUriParam);
+
+            // This part is the where we prepare the response as a OIDC Provider would do.
+            // When returned from the OIDC Provider user will have an Authorization code. This code can have
+            // different formats and this component does not need to understand it. It just exchanged code with
+            // OIDC Provider to get a JWT ID token in response. In this test we create a JWTToken as code to make
+            // the exchange in OidcProviderMock simple
+            string authorizationCode = CreateOidcCode("1337", "1337", nonceParam);
+            string redirectFromOidcProviderUri = GetAuthenticationUrlWithToken(redirectUriParam, "SFSDF", authorizationCode, "https://idprovider.azurewebsites.net/");
+            HttpRequestMessage redirectFromOidcProviderRequest = new HttpRequestMessage(HttpMethod.Get, redirectFromOidcProviderUri);
+
+            // Act 2. This simulates the request the browser will do when user is authenticated at OIDC provider and returns to Altinn authentication.
+            HttpResponseMessage redirectFromOidcProviderResponse = await client.SendAsync(redirectFromOidcProviderRequest);
+
+            // Assert: Now the user should get a bad request since nonce is wrong
```

Comment:
Is it the state or the nonce value that is wrong? From the code it looks like it is the state parameter that is tempered with.

Reply:
fixed text